helix-term/src/commands/dap.rs

```diff
@@ -0,0 +1,517 @@
+use super::{align_view, Align, Context, Editor};
+use crate::ui::{FilePicker, Picker};
+use helix_core::Selection;
+use helix_dap::{self as dap, Client};
+use helix_lsp::block_on;
+
+use serde_json::{to_value, Value};
+use tokio_stream::wrappers::UnboundedReceiverStream;
+
+use std::collections::HashMap;
+
+// general utils:
+pub fn dap_pos_to_pos(doc: &helix_core::Rope, line: usize, column: usize) -> Option<usize> {
+    // 1-indexing to 0 indexing
+    let line = doc.try_line_to_char(line - 1).ok()?;
+    let pos = line + column;
+    // TODO: this is probably utf-16 offsets
+    Some(pos)
+}
+
+pub fn resume_application(debugger: &mut Client) {
+    if let Some(thread_id) = debugger.thread_id {
+        debugger
+            .thread_states
+            .insert(thread_id, "running".to_string());
+        debugger.stack_frames.remove(&thread_id);
+    }
+    debugger.active_frame = None;
+    debugger.thread_id = None;
+}
+
+pub async fn select_thread_id(editor: &mut Editor, thread_id: isize, force: bool) {
+    let debugger = match &mut editor.debugger {
+        Some(debugger) => debugger,
+        None => return,
+    };
+
+    if !force && debugger.thread_id.is_some() {
+        return;
+    }
+
+    debugger.thread_id = Some(thread_id);
+
+    // fetch stack trace
+    // TODO: handle requesting more total frames
+    let (frames, _) = debugger.stack_trace(thread_id).await.unwrap();
```

Comment:
Need to keep calm and not panic here if trying to switch threads while target is running

Reply:
If it's running then there's no debugger.thread_id so it'll exit I the condition check earlier